### PR TITLE
Increase the K tile size in L1 for matmul ops

### PR DESCRIPTION
**Pull request by yzhang93**

### CODE REVIEW

**Comment by yzhang93** [comment]
Currently 5 bf16 ci tests failed because of `[AIE ERROR] _XAie_LoadProgMemSection():231: Overflow of program memory.` All i32 and i8 tests passed.

**Comment by Abhishek-Varma** [comment]
> Currently 5 bf16 ci tests failed because of `[AIE ERROR] _XAie_LoadProgMemSection():231: Overflow of program memory.` All i32 and i8 tests passed.
> 
> @Abhishek-Varma Did you see the similar error? How do you solve it?

Yes, I'm seeing this issue for larger shape Matmul + Truncf bf16.

From what I could speculate yesterday, in Matmul + Truncf's case, is that it seemed to be due to vector unrolling of `arith.truncf`. But in this CI I see that it's happening now for Batch Matmul too.
Trying to play with Peano flags today to see if it helps.

**Comment by Abhishek-Varma** [comment]
Since both of us are facing the same issue - I tried looking into the CI failure (as the current speculation was that `arith.truncf`'s vector unrolling is the culprit - so wanted to check what the delta here is, in case it helps solve both of our cases).

I was trying to compare the e2e IR for the case of PM issue (Program Memory issue) with the one that doesn't use this patch - to see what the delta is.

The only difference (besides the new tile size) that I could see was in the `MAIN` section's `scf.for` total iteration count.
For the case with PM issue the total iteration count is `2` whereas for the one that doesn't has `6`.
And due to this `AMDAIEAcquireReleaseToUseLock`, when trying to unroll, seems to be changing the IR structure from :-
```
core {
    // PROLOGUE
       vector.contract 
    // MAIN
    scf.for 
         vector.contract
    // EPILOGUE
       vector.contract
}
```

To

```
core {
    // PROLOGUE
       vector.contract 
    // MAIN
       vector.contract
       vector.contract
    // EPILOGUE
       vector.contract
}
```

But I don't think this is the cause of PM exhaustion - when trying to understand the same issue for Truncf as well.

I tried changing the Peano flags too :-
```
    -O2 \
    --inline-threshold=10 \
    -S input.ll \
    --disable-builtin=memset \
    -o input.opt.ll \
    -vectorize-loops=false \
    -vectorize-slp=false \
    --two-entry-phi-node-folding-threshold=10 \
    -mandatory-inlining-before-opt=false \
    -basic-aa-full-phi-analysis=true \
    -basic-aa-max-lookup-search-depth=10
```
but none seem to work for both this as well as the Truncf PM issue. :/

**Comment by yzhang93** [comment]
> @jtuyls suggested to add `--disable-loop-unrolling` in the Peano flag and it worked!
> 
> Here is the [e2e IR](https://gist.github.com/Abhishek-Varma/f4e42589480db1c534a5a285afe4cca2) of one of the test's failure with your patch - I verified the output too.
> 
> It also solved the PM issue for Truncf!
> 
> Thanks Jorn!

@Abhishek-Varma Thanks for looking into the issue! So what's the way to move forward? We disable that flag in Peano or add some optimizations?

**Comment by jtuyls** [comment]
> s for looking into the issue! So what's the way to move forward? We disable that flag in Peano or add some optimizations?

Maybe we can check whether it has any performance impact? At some point, this might matter a lot, but not sure at this point. We could make it a an optional flag for now? And potentially, we can unroll/partially unroll on our side?

**Comment by yzhang93** [comment]
> > s for looking into the issue! So what's the way to move forward? We disable that flag in Peano or add some optimizations?
> 
> Maybe we can check whether it has any performance impact? At some point, this might matter a lot, but not sure at this point. We could make it a an optional flag for now? And potentially, we can unroll/partially unroll on our side?

Unfortunately, it does have performance impact. For large size such as 4096x4096x2048, the total execution time doubles while disabling the loop unrolling in peano :(

**Comment by yzhang93** [comment]
@Abhishek-Varma After rebasing on your outlining patches the PM issues are solved. Thanks for that.

However, now all the ukernel tests are failed. Please take a look and let me know if we'd want to keep the original tile sizes for ukernels.

**Comment by Abhishek-Varma** [comment]
> @Abhishek-Varma After rebasing on your outlining patches the PM issues are solved. Thanks for that.
> 
> However, now all the ukernel tests are failed. Please take a look and let me know if we'd want to keep the original tile sizes for ukernels.

@yzhang93 with the `mm.cc` delta in https://github.com/nod-ai/iree-amd-aie/pull/878 (basically added `mm.cc` changes on top of this PR) the compilation failure in the CI is resolved but now the `.vmfb` is generating incorrect results. It'd be nice to have this working with the new L1 sizes itself if it's an issue of incorrect results. CC: @erwei-xilinx could you please help take a look ? Previously we were generating a block of `32 x 32 x 32` (M x N x K), but now we're generating `32 x 32 x 64`. I tried looking into the raw ukernel but the constraints it expects seems to be satisfied.
NOTE: The incorrect result is yielded even without [this commit](https://github.com/nod-ai/iree-amd-aie/pull/878/commits/e56221b41c0958c8de77258c05785ad89124f1b9) on the PR branch I shared.

**Comment by erwei-xilinx** [comment]
> @yzhang93 with the `mm.cc` delta in #878 (basically added `mm.cc` changes on top of this PR) the compilation failure in the CI is resolved but now the `.vmfb` is generating incorrect results. It'd be nice to have this working with the new L1 sizes itself if it's an issue of incorrect results. CC: @erwei-xilinx could you please help take a look ? Previously we were generating a block of `32 x 32 x 32` (M x N x K), but now we're generating `32 x 32 x 64`. I tried looking into the raw ukernel but the constraints it expects seems to be satisfied. NOTE: The incorrect result is yielded even without [this commit](https://github.com/nod-ai/iree-amd-aie/pull/878/commits/e56221b41c0958c8de77258c05785ad89124f1b9) on the PR branch I shared.

I added a comment in https://github.com/nod-ai/iree-amd-aie/pull/878. Don't know if that's the reason for this numerical failure.

**Comment by Abhishek-Varma** [comment]
> > @yzhang93 with the `mm.cc` delta in #878 (basically added `mm.cc` changes on top of this PR) the compilation failure in the CI is resolved but now the `.vmfb` is generating incorrect results. It'd be nice to have this working with the new L1 sizes itself if it's an issue of incorrect results. CC: @erwei-xilinx could you please help take a look ? Previously we were generating a block of `32 x 32 x 32` (M x N x K), but now we're generating `32 x 32 x 64`. I tried looking into the raw ukernel but the constraints it expects seems to be satisfied. NOTE: The incorrect result is yielded even without [this commit](https://github.com/nod-ai/iree-amd-aie/pull/878/commits/e56221b41c0958c8de77258c05785ad89124f1b9) on the PR branch I shared.
> 
> I added a comment in #878. Don't know if that's the reason for this numerical failure.

Thank you @erwei-xilinx ! The numerical issue is resolved.

@yzhang93 can you please cherry pick the `mm.cc` delta from the attached PR link into this PR's branch? This should unblock you.

**Review by Abhishek-Varma — CHANGES_REQUESTED** [review]

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Target/mm.cc [hunk 1]
```diff
@@ -181,7 +181,7 @@ void matmul_vectorized(const T_in *__restrict pA, unsigned offsetA,
           C33.mac(A3, B3);
 
           for (unsigned i = 1; i < colA; ++i)
-            chess_prepare_for_pipelining chess_loop_range(7, ) {
+            chess_prepare_for_pipelining chess_loop_range(3, ) {
```
[the remark this answers is not in the document]
Comment (yzhang93) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Target/mm.cc]: yeah, it works.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/KernelDispatch.cpp [hunk 2]
```diff
@@ -229,7 +229,11 @@ FailureOr<ParameterSetting> ParameterSetting::create(
     // the second level of inner pack size (vector instruction size).
     uint32_t m0Pack = (M0 / 2) % m1Pack == 0 ? (M0 / 2) : M0;
     uint32_t n0Pack = (N0 / 2) % n1Pack == 0 ? (N0 / 2) : N0;
-    uint32_t k0Pack = findLargestFactor(K, maxL1Size);
+
+    // For matmul ops, make the most use of L1 memory by scaling the K tiling
+    // dimension.
```
[the remark this answers is not in the document]
Comment (yzhang93) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/KernelDispatch.cpp]: Added some comments.

**Pull request merged**